### PR TITLE
chore(docker source): upgrade to bollard

### LINKED ISSUE #2947: Replace `shiplift` with `bollard`
This issue is split off of https://github.com/timberio/vector/issues/2118 since that issue is becoming hard to follow.

As part of our work to migrate to `tokio/0.2` and `futures/0.3` we need to upgrade the `docker` source dependencies. The `docker` source uses `shiplift` which is not compatible with the new futures code. Latest versions of `shiplift` appear to be incompatible but `bollard` does not. We should move to `bollard` as a result.

### PR DESCRIPTION
**Pull request by fanatid**
Closes #2947 

Docker source introduced in https://github.com/timberio/vector/pull/787 but shiplift still not support async, so I used [bollard](https://crates.io/crates/bollard).

Bollard produce logs by lines, so we do not need `auto_partial_merge` / `partial_event_marker_field` -- I removed these options.

`futures` updated to `0.3.5` because [take_until](https://docs.rs/futures/0.3.5/futures/stream/trait.StreamExt.html#method.take_until) not available in `0.3.4`. Should I move this update to separate PR?

### CODE REVIEW

File: src/sources/docker.rs [hunk 1]
```diff
@@ -107,15 +94,31 @@ impl SourceConfig for DockerConfig {
...
+            }
+        };
+
+        // Once this ShutdownSignal resolves it will drop DockerSource and by extension it's ShutdownSignal.
+        Ok(Box::new(
+            async move {
+                Ok(tokio::select! {
+                    _ = fut => {}
+                    _ = shutdown.compat() => {}
+                })
```
Comment (fanatid) [inline src/sources/docker.rs:117]: Not sure what's better `tokio::select!` or `futures::select!`?

**Comment by binarylogic** [comment]
Nice work!

> Bollard produce logs by lines, so we do not need auto_partial_merge / partial_event_marker_field -- I removed these options.

Docker splits messages longer than 16kb into multiple messages/events. That's what those options attempted to address. They would automatically merge them back together before being emitted from the source. Does that make sense?

**Comment by fanatid** [comment]
From my understanding `bollard` merge them internally. I'll add test for for checking that this case handled.

**Comment by binarylogic** [comment]
Nice! So we don't break backward compatibility can we silently ignore these options if they are provided and add a note that they are deprecated. We should remove them from the docs though.

**Comment by fanatid** [comment]
I was wrong =( `bollard` not return merged messages and not work correct with logs at all.

- Returned message merge back
- Add 16kb log test
- Submit PR to `bollard` (https://github.com/fussybeaver/bollard/pull/80) and temporary switched to my fork in Cargo.toml

**Comment by binarylogic** [comment]
Bummer, thanks for pushing through this and contributing upstream.

**Comment by fanatid** [comment]
Switched to fixed `bollard`, ready for review...

File: src/sources/docker.rs [hunk 2]
```diff
@@ -263,97 +283,170 @@ impl DockerSource {
...
+                    return;
+                }
 
-                    // Include image check
-                    if let Some(images) = self.esb.core.config.include_images.as_ref() {
-                        let image_check = images.iter().any(|image| &container.image == image);
-                        if !images.is_empty() && !image_check {
-                            continue;
+                // This check is necessary since docker api doesn't have way to include
+                // names into request.
```
Comment (fussybeaver) [inline src/sources/docker.rs:337]: you can actually filter on names and image id's in the container inspect request by setting the `name` and `ancestor` fields to a vector in the `filters` hashmap.

File: src/sources/docker.rs [hunk 3]
```diff
@@ -263,97 +283,170 @@ impl DockerSource {
...
+                                        state.stoped();
+                                    }
+                                }
+                                "start" | "upause" => {
+                                    if let Some(state) = self.containers.get_mut(&id) {
+                                        state.running();
+                                        self.esb.restart(state);
+                                    } else {
+                                        // This check is necessary since docker api doesn't have way to include
+                                        // names into request.
```
Comment (fussybeaver) [inline src/sources/docker.rs:419]: you should be able to use a `filter` hashmap to narrow down calls to the `events` API

File: src/sources/docker.rs (hunk 2 above)
Comment (fanatid) [reply src/sources/docker.rs]: Thanks! I added `ancestor` filter, looks good. Comment about container name is not correct actually, this option works not like match instead `starts_with` applied. I removed wrong comment.

File: src/sources/docker.rs [hunk 4]
```diff
@@ -263,97 +279,164 @@ impl DockerSource {
...
+                            let id = ContainerId::new(id);
+
+                            // Update container status
+                            match action.as_str() {
+                                "die" | "pause" => {
+                                    if let Some(state) = self.containers.get_mut(&id) {
+                                        state.stoped();
+                                    }
+                                }
+                                "start" | "upause" => {
```
Comment (ktff) [inline src/sources/docker.rs:405]: Also `unpause`

File: src/sources/docker.rs [hunk 5]
```diff
@@ -379,256 +462,121 @@ impl DockerSource {
...
-            .and_then(|details| {
-                ContainerMetadata::from_details(&details)
-                    .map_err(|error| error!(message="Metadata extraction failed",%error))
-            });
-
         let this = self.clone();
-        let task = metadata_fetch.and_then(move |metadata| {
-            this.start_event_stream(ContainerLogInfo::new(id, metadata, this.core.now_timestamp))
+        tokio::spawn(async move {
+            Arc::clone(&this.core)
```
Comment (ktff) [inline src/sources/docker.rs:482]: We can just use `this.core.clone()`

File: src/sources/docker.rs [hunk 6]
```diff
@@ -379,256 +462,121 @@ impl DockerSource {
...
+                        partial_event_marker_field.clone(),
+                        auto_partial_merge,
+                        &mut partial_event_merge_state,
+                    )),
+                    Err(error) => {
+                        // On any error, restart connection
+                        match error.kind() {
+                            // http::StatusCode::NOT_IMPLEMENTED
+                            DockerErrorKind::DockerResponseServerError { status_code, .. }
+                                if *status_code == 501 =>
```
Comment (ktff) [inline src/sources/docker.rs:547]: Can we use `StatusCode::NOT_IMPLEMENTED as u32` or something like that so to avoid the need for the above comment.

File: src/sources/docker.rs [hunk 7]
```diff
@@ -379,256 +462,121 @@ impl DockerSource {
...
+
+        let stream = self.core.docker.logs(info.id.as_str(), options);
         info!(
             message = "Started listening logs on docker container",
             id = field::display(info.id.as_str())
         );
 
         // Create event streamer
-        let mut state = Some((self.main_send.clone(), info));
+        let mut info = Some(info);
```
Comment (ktff) [inline src/sources/docker.rs:527]: We can just use info directly without the option thanks to wonderful `async`.

File: src/sources/docker.rs [hunk 8]
```diff
@@ -643,7 +591,7 @@ impl ContainerId {
     }
 
     fn as_str(&self) -> &str {
-        std::str::from_utf8(self.0.borrow()).expect("Bytes should be a still valid String")
+        std::str::from_utf8(&self.0).expect("Bytes should be a still valid String")
```
Comment (ktff) [inline src/sources/docker.rs:594]: Let's correct my pasts sentence
`Container Id Bytes aren't String`

File: src/sources/docker.rs (hunk 1 above)
Comment (fanatid) [reply src/sources/docker.rs:135]: Yeap, `tokio` more about runtime. I wondered is there any other difference? For example `mpsc` faster in `tokio` than in `futures`. Also we already use `tokio::select!`: https://github.com/timberio/vector/blob/01b37785f38159388636ff9954a3675c1ae0d049/src/sinks/file/mod.rs#L118

File: src/sources/docker.rs [hunk 9]
```diff
@@ -263,97 +279,164 @@ impl DockerSource {
...
+    async fn handle_running_containers(mut self) -> crate::Result<Self> {
+        let mut filters = HashMap::new();
+
+        // Apply include filters
+        if let Some(include_labels) = &self.esb.core.config.include_labels {
+            filters.insert("label".to_owned(), include_labels.clone());
+        }
+
+        if let Some(include_images) = &self.esb.core.config.include_images {
+            filters.insert("ancestor".to_owned(), include_images.clone());
```
[the remark this answers is not in the document]
Comment (fanatid) [reply src/sources/docker.rs:309]: I expected too, but `ancestor` works. Probably it's called `ancestor` because container launched on some image and add own layer of file changes, so `ancestor` sounds reasonable.

File: src/sources/docker.rs (hunk 5 above)
Comment (fanatid) [reply src/sources/docker.rs:482]: Yes we can, but I think `Arc::clone` is more clear.

File: src/sources/docker.rs [hunk 10]
```diff
@@ -157,47 +178,42 @@ impl DockerSourceCore {
...
-        options.filter(
-            vec!["start", "upause", "die", "pause"]
-                .into_iter()
-                .map(|s| EventFilter::Event(s.into()))
-                .collect(),
+        filters.insert(
+            "event".to_owned(),
+            vec![
+                "start".to_owned(),
+                "upause".to_owned(),
```
[the remark this answers is not in the document]
Comment (fanatid) [reply src/sources/docker.rs:194]: :scream: it's hard to notice skipping 1 letter  in middle of word

File: src/sources/docker.rs [hunk 11]
```diff
@@ -39,12 +42,10 @@ lazy_static! {
     static ref CONTAINER: Atom = Atom::from("container_id");
 }
 
-type DockerEvent = shiplift::rep::Event;
-
 #[derive(Deserialize, Serialize, Debug, Clone)]
 #[serde(deny_unknown_fields, default)]
 pub struct DockerConfig {
-    include_containers: Option<Vec<String>>,
+    include_containers: Option<Vec<String>>, // Starts with actually, not include
```
[the remark this answers is not in the document]
Comment (fanatid) [reply src/sources/docker.rs:48]: Just added extra comment, because it's possible read code without reading docs, so this make things more obvious.

File: src/sources/docker.rs (hunk 6 above)
Comment (fanatid) [reply src/sources/docker.rs:547]: Changed to constant from `http` crate :+1:

File: src/sources/docker.rs (hunk 7 above)
Comment (fanatid) [reply src/sources/docker.rs:527]: Indeed, with current code we can, thanks! I was need `Option<T>` for some code in middle of changes.

File: src/sources/docker.rs [hunk 12]
```diff
@@ -157,47 +178,42 @@ impl DockerSourceCore {
...
         }
 
-        options.filter(filters);
-
-        self.docker.events(&options.build())
+        self.docker.events(Some(EventsOptions {
+            since: self.now_timestamp.clone(),
+            // Handler in Docker API:
+            // https://github.com/moby/moby/blob/c833222d54c00d64a0fc44c561a5973ecd414053/api/server/router/system/system_routes.go#L155
+            until: MAX_DATE.and_time(NaiveTime::from_hms(0, 0, 0)).unwrap(),
```
[the remark this answers is not in the document]
Comment (fanatid) [reply src/sources/docker.rs:214]: Then docker daemon will not apply this to filter. Question how create `DateTime` which will be serialized to zero?

[the remark this answers is not in the document]
Comment (fussybeaver) [reply src/sources/docker.rs:214]: I don't think that is possible in bollard at the moment (I guess it needs to serialize `until` to a `-1`?), maybe it can be added in future.

File: src/sources/docker.rs (hunk 1 above)
Comment (ktff) [reply src/sources/docker.rs:135]: Not that I know of. It should be a small difference in both cases. With what's said here, retaining `tokio::select` seems fine.

File: src/sources/docker.rs (hunk 5 above)
Comment (ktff) [reply src/sources/docker.rs:482]: Is the information that we are cloning `Arc` and doing it by borrowing `this.core` relevant here? If not, `this.core.clone()` has less noise.

**Comment by fanatid** [comment]
@ktff thank you for review!

**Comment by binarylogic** [comment]
Nice work @fanatid, I know this wasn't an easy one.
